File: razarion-share/src/main/java/com/btxtech/shared/utils/SignalGenerator.java
package com.btxtech.shared.utils;

/**
 * Created by Beat
 * on 12.01.2019.
 */
public interface SignalGenerator {

    static double triangle(long millis, int durationMs, int offsetMs) {
        long totMillis = millis + offsetMs;
        int mod = (int) (totMillis % durationMs);
        if (mod < durationMs / 2) {
            // Raising
            return 2.0 * mod / durationMs;
        } else {
            // Falling
            return 2.0 - 2.0 * mod / durationMs;
        }
    }

    static double sawtooth(long millis, int durationMs, int offsetMs) {
        if(durationMs == 0.0) {
            return 0;
        }
        long totMillis = millis + offsetMs;
        return (double) (totMillis % durationMs) / (double)durationMs; // Saegezahn
    }

    static double sinus(long millis, int durationMs, int offsetMs) {
        if(durationMs == 0.0) {
            return 0;
        }
        return (1.0 + Math.sin(((millis % durationMs) / (double) durationMs + ((double) offsetMs / (double) durationMs)) * MathHelper.ONE_RADIANT)) / 2.0;
    }


}

File: razarion-server/src/main/angular/frontend/src/app/game/signal-generator.ts
export class SignalGenerator {
  static triangle(millis: number, durationMs: number, offsetMs: number): number {
    let totMillis = millis + offsetMs;
    let mod = totMillis % durationMs;
    if (mod < durationMs / 2) {
      // Raising
      return 2.0 * mod / durationMs;
    } else {
      // Falling
      return 2.0 - 2.0 * mod / durationMs;
    }
  }

  static sawtooth(millis: number, durationMs: number, offsetMs: number): number {
    if (durationMs == 0.0) {
      return 0;
    }
    let totMillis = millis + offsetMs;
    return (totMillis % durationMs) / durationMs; // Saegezahn
  }
}
